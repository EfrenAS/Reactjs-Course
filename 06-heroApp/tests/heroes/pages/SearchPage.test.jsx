import { cleanup, fireEvent, render, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router'
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest'
import SearchPage from '../../../src/heroes/pages/SearchPage'

describe('<SearchPage /> tests', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  afterEach(() => {
    cleanup()
  })

  vi.mock(import('react-router'), async (importOriginal) => {
    const actual = await importOriginal()
    return {
      ...actual,
      navigate: vi.fn()
    }
  })

  test('should render the search page', () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    )

    expect(container).toMatchSnapshot()
  })

  test('Should render the search page with a Batman and the input field with the value of queryString', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchPage />
      </MemoryRouter>
    )

    const input = screen.getByRole('textbox')
    const img = screen.getByRole('img')

    expect(input.value).toBe('batman')
    expect(img.src).toContain('/assets/dc-batman-min.jpg')
    expect(screen.queryByText('Find to favorite Hero')).toBeFalsy()
  })

  test('Should render the message "No heroes found" when the search is empty or not found any hero', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <SearchPage />
      </MemoryRouter>
    )

    expect(screen.getByText('No heroes found')).toBeTruthy()
  })

  test('Should call navigate when the form is submitted', () => {
    render(
      <MemoryRouter initialEntries={['/search']}>
        <SearchPage />
      </MemoryRouter>
    )

    const input = screen.getByRole('textbox')
    const form = screen.getByRole('form')

    fireEvent.change(input, { target: { name: 'searchHero', value: 'batman' } })
    fireEvent.submit(form)

    expect(vi.fn()).toHaveBeenCalledWith('?q=batman', { viewTransition: true })
  })
})
