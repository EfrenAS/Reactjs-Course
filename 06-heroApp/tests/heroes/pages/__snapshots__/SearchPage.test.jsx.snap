// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SearchPage /> tests > should render the search page 1`] = `
<div>
  <main
    class="md:w-2/3 h-[calc(100vh-5rem)] mx-auto"
  >
    <section
      aria-labelledby="search-hero"
      class="p-2"
    >
      <header
        class="flex flex-col gap-2 items-center"
      >
        <h1
          class="text-center uppercase text-xl font-bold text-amber-500"
        >
          Find your favorite hero
        </h1>
        <form
          aria-label="form"
          class="flex align-center gap-8 p-5 w-full justify-evenly flex-col sm:flex-row @container"
        >
          <input
            aria-label="search hero"
            class="border-b-1 border-gray-500 p-2 focus:outline-none focus:border-amber-500 w-full @xs:w-7/12"
            name="searchHero"
            placeholder="batman, superman, spiderman, ironman..."
            type="text"
            value=""
          />
          <button
            class="bg-amber-500/50 px-1 py-0.5 rounded-xs hover:bg-amber-600 transition-all duration-300 ease-in-out cursor-pointer"
          >
            Find hero
          </button>
        </form>
      </header>
    </section>
    <section
      aria-labelledby="hero-info"
      class="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 xl:grid-cols-4 gap-6 p-4"
    >
      <p
        aria-labelledby="message-find-hero"
        class="text-center text-gray-500"
      >
        Find to favorite Hero
      </p>
    </section>
  </main>
</div>
`;
